transport/src/main/java/io/netty/channel/pool/FixedChannelPool.java

```diff
@@ -236,48 +236,57 @@ public void run() {
                 });
             }
         } catch (Throwable cause) {
-            promise.setFailure(cause);
+            promise.tryFailure(cause);
         }
         return promise;
     }
 
     private void acquire0(final Promise<Channel> promise) {
-        assert executor.inEventLoop();
-
-        if (closed) {
-            promise.setFailure(new IllegalStateException("FixedChannelPool was closed"));
-            return;
-        }
-        if (acquiredChannelCount.get() < maxConnections) {
-            assert acquiredChannelCount.get() >= 0;
+        try {
+            assert executor.inEventLoop();
 
-            // We need to create a new promise as we need to ensure the AcquireListener runs in the correct
-            // EventLoop
-            Promise<Channel> p = executor.newPromise();
-            AcquireListener l = new AcquireListener(promise);
-            l.acquired();
-            p.addListener(l);
-            super.acquire(p);
-        } else {
-            if (pendingAcquireCount >= maxPendingAcquires) {
-                tooManyOutstanding(promise);
+            if (closed) {
+                promise.setFailure(new IllegalStateException("FixedChannelPool was closed"));
+                return;
+            }
+            if (acquiredChannelCount.get() < maxConnections) {
+                assert acquiredChannelCount.get() >= 0;
+
+                // We need to create a new promise as we need to ensure the AcquireListener runs in the correct
+                // EventLoop
+                Promise<Channel> p = executor.newPromise();
+                AcquireListener l = createAcquireListener(promise);
```

Comment:
Since we are calling `super.acquire()` below, we can utilize the same mode as testing `SimpleChannelPool` here by implementing the `ChannelPoolHandler`. I would prefer to avoid expanding API surface area just for tests.

Reply:
I was thinking like that but then I realized that `super.acquire()` has a `try/catch`. In this case, the failure won't be caught by `FixedChannelPool`. The only way that I found was changing the API.

Reply:
I think testing the behavior that a custom handler throwing, correctly completes the promise is more important to test than to test whether `FixedChannelPool` has a `try-finally` block or not. So, even though `super.acquire()` is making sure promise is completed correctly, it is still a legitimate test for `FixedChannelPool`.

Reply:
@NiteshKant Hi, before implementing the same test case in `FixedChannelPoolTest`, could you please review again the test in `SimpleChannelPoolTest`
It should be like ( I haven't done the commit yet )
```java
@Test
    public void testChannelAcquiredException() throws InterruptedException, TimeoutException, ExecutionException {
        final LocalAddress addr = new LocalAddress(getLocalAddrId());
        final EventLoopGroup group = new DefaultEventLoopGroup();

        // Start server
        final ServerBootstrap sb = new ServerBootstrap()
              .group(group)
              .channel(LocalServerChannel.class)
              .childHandler(new ChannelInitializer<LocalChannel>() {
                  @Override
                  protected void initChannel(LocalChannel ch) throws Exception {
                      ch.pipeline().addLast(new ChannelInboundHandlerAdapter());
                  }
              });
        final Channel sc = sb.bind(addr).syncUninterruptibly().channel();

        // Create pool, acquire and return channels
        final Bootstrap bootstrap = new Bootstrap()
              .channel(LocalChannel.class).group(group).remoteAddress(addr);
        final NullPointerException exception = new NullPointerException();
        final SimpleChannelPool pool = new SimpleChannelPool(bootstrap, new ChannelPoolHandler() {
            @Override
            public void channelReleased(Channel ch) throws Exception {
            }
            @Override
            public void channelAcquired(Channel ch) throws Exception {
                throw exception;
            }
            @Override
            public void channelCreated(Channel ch) throws Exception {
            }
        });

        final AtomicBoolean run = new AtomicBoolean();
        Promise<Channel> channelPromise = bootstrap.config().group().next().<Channel>newPromise();
        channelPromise.addListener(new GenericFutureListener<Future<Channel>>() {
            @Override
            public void operationComplete(Future<Channel> future) {
                assertTrue(future.cause() == exception);
                run.set(true);
            }
        });

        try {
            pool.acquire(channelPromise).get();
        } catch (ExecutionException e) {
            assertTrue(e.getCause() == exception);
        }

        assertTrue(run.get());

        sc.close().sync();
        pool.close();
        group.shutdownGracefully();
    }
```

I have added the `AtomicBoolean` because if the assert fails for some reason, we will see only the stack trace and the test won't fail.

WDYT?

Reply:
Aah ... ok you can probably do this instead of the `AtomicBoolean` and the listener:

```java
        final Future<Channel> futureChannel = pool.acquire();
        try {
            futureChannel.sync();
        } catch (NullPointerException cause) {
            assertSame(exception, cause);
        }
```

Also, from the `ChannelPoolHandler` anonymous implementation here, remove the redundant `throws Exception` clauses.

Reply:
Done